lessons/lesson3.md

```diff
@@ -0,0 +1,151 @@
+    };
+
+    return returnObject
+  } catch (error) {
+    myAppLogger.error(error)
+    return {
+      statusCode: 500,
+      body: { error: 'server error!' }
+    }
+  }
+}
+
+exports.main = main
+```
+What happens here is that the action exposes a `main` function, which accepts a list of params from the client. It checks
+that required params for using the `EventsSDK`. 
+
+Next, let's see how the web UI communicates with the backend. All web assets are placed in the web-src folder.
+Beside a few auto-generated files that are useful for running your app on Adobe Experience Cloud (ExC) Shell, App.js is the extension point of your UI.
+By default, it contains a form that lists all available backend actions, allows you to select the action to be invoke
+```javascript
+render () {
+  return (
+    // ErrorBoundary wraps child components to handle eventual rendering errors
+    <ErrorBoundary onError={this.onError} FallbackComponent={this.fallbackComponent} >
+    <section class="banner style1 orient-left content-align-left image-position-right fullscreen onload-image-fade-in onload-content-fade-right">
+        <div class="content">
+            <h1>Basel</h1>
+              <p class="major">A city in northwestern Switzerland on the river Rhine. Basel is Switzerland's third-most-populous city. The city is known for its many internationally renowned museums, ranging from the Kunstmuseum, the first collection of art accessible to the public in Europe (1661) and the largest museum of art in the whole of Switzerland.</p>
+              <ul class="action">
+                <li><a href={actions.event}>Like</a></li>
+              </ul>
+        </div>
+        <div class="image">
+            <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f5/Basel_-_M%C3%BCnsterpfalz1.jpg/2560px-Basel_-_M%C3%BCnsterpfalz1.jpg" alt="" />
+        </div>
+      </section>
+    </ErrorBoundary>
+  )
+  }
+}
+
+```
+source code is [here](https://github.com/AdobeDocs/adobeio-codelab-customevent-demo/blob/master/web-src/src/App.js)
+
+### Run and deploy the Firefly App
+You can run the firefly app locally by execute the below command with AIO CLI
+```bash
+aio app run
+```
+This command will deploy the `event` action onto I/O Runtime, and spins up a local instance for the UI. When the app is up and running, it can be seen at https://localhost:9080. You should be able to see the UI of the app
+it is also possible to access the app from ExC Shell: https://experience.adobe.com/?devMode=true#/apps/?localDevUrl=https://localhost:9080. You might be asked to log in using your Adobe ID.  When opened, the UI is almost similar to that when you access it on localhost, except the ExC Shell on top.
+
+Once you are satisfied with your app, you can deploy your app using the `namespace` and `auth` you get from console integration, execute the below command with AIO CLI
+```bash
+aio app deploy
+```
+This command will deploy to your namespace, you will get the URL like 
```

Comment:
please check the markdown result in Preview mode. The URL is not displayed properly.

Reply:
fixed